File: Striver/Day 2 : Array 2/Merge Intervals.cpp
//Question: https://leetcode.com/problems/merge-intervals/

//Solution:
vector<vector<int>> merge(vector<vector<int>>& intervals) {
    vector<vector<int>> ans;
    sort(intervals.begin(), intervals.end()); //sort acc to first element
    int first = intervals[0][0], sec = intervals[0][1];
    for(int i=0;i<intervals.size(); i++) {
        if(intervals[i][0] <= sec) {
            sec = max(sec, intervals[i][1]);      //if intervals overlap : merge them
        } else {
            ans.push_back({first, sec});      //add merged interval(from previous iteration) to the solution
            first = intervals[i][0];
            sec = intervals[i][1];
        }
    }
    ans.push_back({first, sec});      //the last one interval is left out. So, add it after the loop ends
    return ans;
}

File: Striver/Day 2 : Array 2/Merge Intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        
     Arrays.sort(intervals, (a, b) -> Integer.compare(a[0],b[0]));
        
  
        ArrayList<ArrayList<Integer>> ans = new ArrayList<>();
  
        ArrayList<Integer> temp = new ArrayList<>();
        temp.add(intervals[0][0]);
        temp.add(intervals[0][1]);
        
        ans.add(temp);
        
        
        for(int[] interval : intervals)
        {
            temp = new ArrayList<>();
            
            int pfirst = ans.get(ans.size()-1).get(0);
            int plast  = ans.get(ans.size()-1).get(1);
            int cfirst = interval[0];
            int clast  = interval[1];
            
            if(plast >= cfirst)
            {   
                ans.remove(ans.get(ans.size()-1));
                
                temp.add(pfirst);
                temp.add(Math.max(clast , plast));
                
                ans.add(temp);
            } 
            
            else
            {       
             temp.add(cfirst);
             temp.add(clast);
         
             ans.add(temp);   
            }
     
        }
        
        int[][] result = new int[ans.size()][2];
        

        for(int i=0 ; i<ans.size() ; i++)
        {
            for(int j=0 ; j<ans.get(0).size() ; j++)
            {
                result[i][j] = ans.get(i).get(j);
            }
        }
        
        return result;
    }
    
   
    
}
